main/at_transport_sdio.cpp

```diff
@@ -0,0 +1,390 @@
+    std::lock_guard<std::recursive_mutex> lock(txMutex_);
+    size_t consumable = std::min<size_t>(txBuf_.consumable(), CONFIG_AT_SDIO_BLOCK_SIZE);
+    if (transmitting_ || consumable == 0) {
+        return 0;
+    }
+
+    size_t align = (consumable % 4);
+    if (align) {
+        align = sizeof(uint32_t) - align;
+
+        if (txBuf_.space() < align) {
+            return RESULT_NO_MEMORY;
+        }
+
+        // Dummy write to enforce alignment
+        CHECK(txBuf_.put(nullptr, align));
+        consumable = txBuf_.consumable();
+    }
+
+    if (consumable % 4 != 0) {
+        return RESULT_ERROR;
+    }
+
+    auto ptr = txBuf_.consume(consumable);
+    assert(esp_ptr_dma_capable(ptr));
+
+    esp_err_t ret = sdio_slave_send_queue((uint8_t*)ptr, consumable - align, (void*)consumable, 0);
+    if (ret != ESP_OK) {
+        // Cancel
+        txBuf_.consumeCommit(0, consumable);
+    }
+    CHECK_ESP(ret);
+
+    transmitting_ = true;
+
+    return 0;
+}
+
+int AtSdioTransport::waitTransmissionFinished(unsigned int timeoutMsec) {
+    size_t consume = 0;
+    auto ret = sdio_slave_send_get_finished((void**)&consume, timeoutMsec / portTICK_PERIOD_MS);
+    if (ret == ESP_ERR_TIMEOUT) {
+        return RESULT_TIMEOUT;
+    }
+    CHECK_ESP(ret);
+    {
+        std::lock_guard<std::recursive_mutex> lock(txMutex_);
+        txBuf_.consumeCommit(consume);
+        transmitting_ = false;
+    }
+    return 0;
+}
+
+int AtSdioTransport::getDataLength() const {
+    return (int)rxData_;
+}
+
+int AtSdioTransport::waitWriteComplete(unsigned int timeoutMsec) {
+    // FIXME: busy-ish loop
+    auto start = util::millis();
+    while (util::millis() - start < timeoutMsec) {
+        while(transmitting_) {
+            vTaskDelay(1 / portTICK_PERIOD_MS);
+        }
+        {
+            std::lock_guard<std::recursive_mutex> lock(txMutex_);
+            if (txBuf_.empty()) {
+                return 0;
+            }
+        }
+    }
+    return RESULT_TIMEOUT;
+}
+
+int AtSdioTransport::statusChanged(esp_at_status_type status) {
+    /* Nothing to do here, because we don't use any commands that switch to non-AT mode */
+    return 0;
+}
+
+int AtSdioTransport::preDeepSleep() {
+    return destroyTransport();
+}
+
+int AtSdioTransport::preRestart() {
+    return destroyTransport();
+}
+
+void AtSdioTransport::rxRun() {
+    LOG(INFO, "SDIO transport RX thread started");
+
+    while (!exit_) {
+        // receive data from SDIO host
+        auto ret = fetchData(AT_SDIO_WAKE_UP_PERIOD_MS);
+
+        // notify length to AT core
+        if (ret > 0) {
+            rxData_ += ret;
+            notifyReceivedData(ret, AT_SDIO_WAKE_UP_PERIOD_MS);
+        }
+    }
+
+    LOG(INFO, "SDIO transport RX thread exiting");
+
+    exit_--;
+}
+
+void AtSdioTransport::txRun() {
+    LOG(INFO, "SDIO transport TX thread started");
+
+    while (!exit_) {
+        if (!waitTransmissionFinished(AT_SDIO_WAKE_UP_PERIOD_MS)) {
+            startTransmission();
```

Comment:
What if `destroyTransport()` is called at somewhere right after `startTransmission()` , `waitTransmissionFinished()` will not get executed. Then `destroyTransport()->waitWriteComplete(portMAX_DELAY)` looks make no sense.

Reply:
True, doesn't make sense because we are already `transmitting_ = false`. We can probably leave it as-is, or force run `waitTransmissionFinished()` in `destroyTransport()`.